docker/docker_client.go

```diff
@@ -244,37 +248,96 @@ func newDockerClient(sys *types.SystemContext, registry, reference string) (*doc
 	if err := tlsclientconfig.SetupCertificates(certDir, tr.TLSClientConfig); err != nil {
 		return nil, err
 	}
+	tr.TLSClientConfig.InsecureSkipVerify = sysContextSkipVerify(sys)
+
+	// Find the registry
+	reg, err := sysregistriesv2.FindRegistry(sys, reference)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error loading registries")
+	}
+	if reg != nil {
+		// Retrieve the docker client from the found registry
+		return dockerClientFromSysRegistry(ctx, sys, reg, tr), nil
+	}
 
+	// No registry found within the configuration, fallback to the provided
+	// one
+	return &dockerClient{
+		registry:              registry,
+		sys:                   sys,
+		client:                &http.Client{Transport: tr},
+		insecureSkipTLSVerify: tr.TLSClientConfig.InsecureSkipVerify,
+	}, nil
+}
+
+func sysContextSkipVerify(sys *types.SystemContext) bool {
 	// Check if TLS verification shall be skipped (default=false) which can
 	// either be specified in the sysregistriesv2 configuration or via the
 	// SystemContext, whereas the SystemContext is prioritized.
-	skipVerify := false
 	if sys != nil && sys.DockerInsecureSkipTLSVerify != types.OptionalBoolUndefined {
 		// Only use the SystemContext if the actual value is defined.
-		skipVerify = sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
-	} else {
-		reg, err := sysregistriesv2.FindRegistry(sys, reference)
-		if err != nil {
-			return nil, errors.Wrapf(err, "error loading registries")
+		return sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
+	}
+	return false
+}
+
+// dockerClientFromSysRegistry returns a `*dockerClient` for the provided
+// sysregistriesv2. If the registry contains any mirrors, it will be tried
+// to detect their properties. The mirror will be used if the property
+// detection succeeds. If not, the fallback registry URL will be used instead.
+func dockerClientFromSysRegistry(
+	ctx context.Context,
+	sys *types.SystemContext,
+	registry *sysregistriesv2.Registry,
+	transport *http.Transport,
+) *dockerClient {
+	// registry != nil should already be checked by referring functions
+	// Check for available mirrors
+	for _, mirror := range registry.Mirrors {
+		// Adapt the InsecureSkipVerify only when not already set from the
+		// system context
+		if !transport.TLSClientConfig.InsecureSkipVerify {
+			transport.TLSClientConfig.InsecureSkipVerify = mirror.Insecure
+		}
+
+		client := &dockerClient{
+			client:                &http.Client{Transport: transport},
+			insecureSkipTLSVerify: mirror.Insecure,
+			registry:              mirror.URL,
+			sys:                   sys,
 		}
-		if reg != nil {
-			skipVerify = reg.Insecure
+
+		// Test the registries availability
+		if client.detectProperties(ctx) != nil {
```

Comment:
I’m not sure this is long-term the right place for mirror detection: I would expect mirror servers, if someone configures them at all, to be on-line most of the time.

It would probably be more useful to make a mirror choice based on whether a _specific image_ is available (e.g. the mirror may be configured for all of `docker.io` but only contain known frequently-used images), or maybe even a specific blob (when mirroring is in progress and half of the image blobs are now available) — and given how `GetManifest` etc. work, the latter might be easier to do anyway.

A choice at this point does not _hurt_ and if this is all you need, sure, this PR can do a choice at this point and the choice can perhaps be moved elsewhere later.

Reply:
In general I agree, this is not the most flexible solution. I will play around with it and see what I can do here too. I'm thinking about storing the mirrors at the `dockerClient` and processing them afterwards.

Reply:
That's a very critical part of the mirror (and prefix) semantics. As @mtrmac mentioned, checking if the mirror is available is not sufficient as the *mirror* might not serve as a pull-through cache but as a real registry.

When checking the availability of a given image, we need to rewrite the reference based on the prefix.

Reply:
> When checking the availability of a given image, we need to rewrite the reference based on the prefix.

Oooooops; how have I not noticed that?!

Reply:
This point should be solved with the new approach.